File: as/scanner.c
#include "scanner.h"
#include "../shared/disasm.h"
#include "../shared/intern.h"
#include "../shared/vector.h"
#include <ctype.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>

#define scanner_debug(...)   \
    if (scan->debug) {       \
        printf(__VA_ARGS__); \
    }

#define scanner_error(text) \
    printf("\e[31merror\e[0m :: %s(%zu:%zu): %s\n", scan->file_name, scan->tok->lno, scan->tok->col, text)

#define scanner_warning(text) \
    printf("warning :: %s(%zu:%zu): %s\n", scan->file_name, scan->tok->lno, scan->tok->col, text)

#define scanner_errorf(fmt, ...)                                                                                      \
    {                                                                                                                 \
        char*       sfmt;                                                                                             \
        static char buf[1024];                                                                                        \
        asprintf(&sfmt, "\e[31merror\e[0m :: %s(%zu:%zu): %s", scan->file_name, scan->tok->lno, scan->tok->col, fmt); \
        snprintf(buf, sizeof(buf), sfmt, ##__VA_ARGS__);                                                              \
        printf("%s\n", buf);                                                                                          \
        free(sfmt);                                                                                                   \
    }

scanner_t* scanner_init(char* file_name)
{
    scanner_t* scan = malloc(sizeof(*scan));
    if (scan == NULL) {
        perror("scanner_init");
        exit(-1);
    }

    scan->file_name = file_name;
    scan->file = fopen(scan->file_name, "r");

    if (scan->file == NULL) {
        perror("scanner");
        exit(-1);
    }

    fseek(scan->file, 0, SEEK_END);
    scan->file_size = ftell(scan->file);
    fseek(scan->file, 0, SEEK_SET);

    scan->file_bytes = malloc(scan->file_size);
    if (scan->file_bytes == NULL) {
        perror("scanner");
        exit(-1);
    }

    int    c;
    size_t pos = 0;
    while ((c = fgetc(scan->file)) != EOF) {
        scan->file_bytes[pos++] = (char)c;
    }

    scan->file_bytes[pos] = '\0';
    fseek(scan->file, 0, SEEK_SET);

    scan->pos = 0;
    scan->lno = 1;
    scan->col = 1;

    return scan;
}

void scanner_set_debug(scanner_t* scan, int debug)
{
    scan->debug = debug;
}

void scanner_delete(scanner_t* scan)
{
    fclose(scan->file);
}

#define pos() __pos(scan)
static char* __pos(scanner_t* scan)
{
    return &scan->file_bytes[scan->pos];
}

#define chr() __chr(scan)
static int __chr(scanner_t* scan)
{
    return *pos();
}

#define fwd() __fwd(scan)
static void __fwd(scanner_t* scan)
{
    fgetc(scan->file);
    scan->pos++;
    scan->col++;

    if (scan->tok != NULL) {
        scan->tok->end++;
    }

    if (chr() == '\n') {
        scan->lno++;
        scan->col = 0;
    }
}

#define match(c) __match(scan, (c))
static int __match(scanner_t* scan, int c)
{
    if (chr() == c) {
        fwd();
        return 1;
    }

    return 0;
}

#define fmatch(f) __fmatch(scan, (f))
static int __fmatch(scanner_t* scan, int (*func)(int))
{
    if (func(chr())) {
        fwd();
        return 1;
    }

    return 0;
}

#define expect(c) __expect(scan, c)
static int __expect(scanner_t* scan, int c)
{
    if (!match(c)) {
        scanner_errorf("expected '%c', got '%c'", c, chr());
        return 0;
    }

    return 1;
}

static int is_name_start(int c)
{
    return isalpha(c) || c == '_' || c == '.';
}

static int is_name_part(int c)
{
    return isalpha(c) || isdigit(c) || c == '_' || c == '.';
}

static void make_tok(scanner_t* scan)
{
    scan->tok = malloc(sizeof(*scan->tok));
    *scan->tok = (tok_t){
        .lno = scan->lno,
        .col = scan->col,
        .start = pos(),
        .end = pos()
    };
}

static void scan_name(scanner_t* scan)
{
    scan->tok->kind = TOK_NAME;
    for (; is_name_part(chr());) {
        fwd();
    }

    if (scan->debug) {
        printf("scanner_tok -> %s\n", tok_str(scan->tok));
    }
}

static int char_ival[] = {
    ['0'] = 0,
    ['1'] = 1,
    ['2'] = 2,
    ['3'] = 3,
    ['4'] = 4,
    ['5'] = 5,
    ['6'] = 6,
    ['7'] = 7,
    ['8'] = 8,
    ['9'] = 9,
    ['a'] = 10,
    ['A'] = 10,
    ['b'] = 11,
    ['B'] = 11,
    ['c'] = 12,
    ['C'] = 12,
    ['d'] = 13,
    ['D'] = 13,
    ['e'] = 14,
    ['E'] = 14,
    ['f'] = 15,
    ['F'] = 15,
};

static void scan_num(scanner_t* scan)
{
    scan->tok->kind = TOK_INT;
    size_t len = 0;

    int base = 10;
    if (match('0')) {
        if (match('x') || match('X')) {
            base = 16;
        } else if (match('b') || match('B')) {
            base = 2;
        } else {
            base = 8;
        }
    }

    for (;;) {
        if (match('_')) {
            continue;
        }

        int digit = char_ival[(unsigned char)chr()];

        if (digit == 0 && chr() != '0') {
            break;
        }

        if (digit >= base) {
            scanner_errorf("digit '%c' out of range for base %d", chr(), base);
            digit = 0;
        }

        if (scan->tok->int_val > (ULLONG_MAX - digit) / base) {
            scanner_errorf("integer literal overflow (%zu)", scan->tok->int_val);
            while (fmatch(isdigit)) {
            }

            scan->tok->int_val = 0;
            break;
        }
        scan->tok->int_val = scan->tok->int_val * base + digit;
        fwd();
    }

    scanner_debug("  -> %s\n", tok_str(scan->tok));
}

static void scan_reg(scanner_t* scan)
{
    scan->tok->kind = TOK_REG;
    for (; is_name_part(chr());) {
        fwd();
    }

    char* reg_str = intern_range(scan->tok->start, scan->tok->end);
    int   reg = 1 + reg_index(reg_str);

    if (reg == -1) {
        scanner_errorf("unknown register '%s'", reg_str);
    }

    scan->tok->int_val = reg - 1;

    scanner_debug("  -> %s\n", tok_str(scan->tok));
}

static void discard_spaces(scanner_t* scan)
{
    while (fmatch(isspace)) {
    }
}

static int comment(scanner_t* scan)
{
    int b = 0;
    discard_spaces(scan);
    if (match(';')) {
        while (!match('\n')) {
            fwd();
        }
        b = 1;
    }
    return b;
}

static void mkstr(tok_t* tok)
{
    char*  str = intern_range(tok->start + 1, tok->end - 1);
    size_t len = strlen(str);

    tok->str_val = malloc(len);
    if (tok->str_val == NULL) {
        perror("mkstr");
        exit(-1);
    }

    char* rc = tok->str_val;
    for (char* c = str; c != '\0';) {
        switch (*c) {
        case '\\':
            switch (*++c) {
            case '0':
                *rc++ = '\0';
                break;

            case 'n':
                *rc++ = '\n';
                break;

            case 'r':
                *rc++ = '\r';
                break;

            case 't':
                *rc++ = '\t';
                break;

            case '\\':
                *rc++ = '\\';
                break;

            case '"':
                *rc++ = '"';
                break;

            case '\'':
                *rc++ = '\'';
                break;
            }
            break;

        default:
            *rc++ = *c++;
        }
    }
}

static void scan_escape(scanner_t* scan)
{
    expect('\\');
    switch (chr()) {
    case '0':
        scan->tok->int_val = '\0';
        break;

    case 'n':
        scan->tok->int_val = '\n';
        break;

    case 'r':
        scan->tok->int_val = '\r';
        break;

    case 't':
        scan->tok->int_val = '\t';
        break;

    case '\\':
        scan->tok->int_val = '\\';
        break;

    case '\'':
        scan->tok->int_val = '\'';
        break;

    case '"':
        scan->tok->int_val = '"';
        break;

    default:
        scanner_errorf("unknown escape sequence '\\%c'", chr());
    }

    fwd();
}

static void scan_string(scanner_t* scan)
{
    scan->tok->kind = TOK_STR;
    for (; !feof(scan->file);) {
        if (match('\n')) {
            scanner_error("unclosed string literal");
            return;
        }

        if (fmatch(iscntrl)) {
            scanner_error("illegal control character in string literal");
            continue;
        }

        if (match('"')) {
            return;
        }

        if (match('\\')) {
            scan_escape(scan);
        }

        fwd();
    }
    mkstr(scan->tok);
    scanner_debug("  -> %s\n", tok_str(scan->tok));
}

static void scan_char(scanner_t* scan)
{
    scan->tok->kind = TOK_INT;
    switch (chr()) {
    case '\\':
        scan_escape(scan);
        break;
    default:
        scan->tok->int_val = chr();
        fwd();
    }
    expect('\'');
    scanner_debug("  -> %s\n", tok_str(scan->tok));
}

void scanner_tok(scanner_t* scan)
{
    scan->tok = NULL;

    while (comment(scan)) {
    }

    make_tok(scan);

    if (feof(scan->file)) {
        scan->tok->kind = TOK_EOF;
        return;
    }

    if (fmatch(is_name_start)) {
        scan_name(scan);
        return;
    }

    if (isdigit(chr())) {
        scan_num(scan);
        return;
    }

    if (match('%')) {
        scan_reg(scan);
        return;
    }

    if (match('"')) {
        scan_string(scan);
        return;
    }

    if (match('\'')) {
        scan_char(scan);
        return;
    }

    scan->tok->kind = chr();
    fwd();
}

File: src/asm/scanner.go
package asm

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"strconv"
	"unicode"
	"unicode/utf8"
)

type Pos struct {
	Line, Col int
}

func (p Pos) String() string {
	return fmt.Sprintf("(%d, %d)", p.Line, p.Col)
}

type Token struct {
	File   *File
	Pos    Pos
	Kind   int
	Text   string
	IntVal uint64
	StrVal string
}

const (
	TokEOF = 0
	TokReg = 128 + iota
	TokDirective
	TokName
	TokInt
	TokString
)

func kindName(kind int) string {
	switch kind {
	case TokReg:
		return "reg"

	case TokDirective:
		return "directive"

	case TokName:
		return "name"

	case TokInt:
		return "int"

	case TokString:
		return "string"

	default:
		return string(kind)
	}
}

func (t *Token) String() (str string) {
	switch {
	case t.Kind == TokEOF:
		str = fmt.Sprintf("token(eof) %s", t.Pos)

	case t.Kind == TokInt:
		str = fmt.Sprintf("token(int) %s '%s' %d", t.Pos, t.Text, t.IntVal)

	case t.Kind > 127:
		str = fmt.Sprintf("token(%s) %s '%.*s'", kindName(t.Kind), t.Pos, 32, t.Text)

	default:
		str = fmt.Sprintf("token('%s') %s", kindName(t.Kind), t.Pos)
	}
	return
}

type Scanner struct {
	File  *File
	Data  []rune
	pos   int
	Pos   Pos
	Token *Token
}

func Scan(file *File) *Scanner {
	bytes, err := ioutil.ReadFile(file.Path)
	if err != nil {
		fmt.Printf("error: %s\n", err.Error())
		os.Exit(1)
	}

	return &Scanner{
		File:  file,
		Data:  []rune(string(bytes)),
		pos:   0,
		Pos:   Pos{1, 1},
		Token: nil,
	}
}

func (s *Scanner) errorf(msg string, args ...interface{}) {
	s.File.Errorf(s.Pos, msg, args...)
}

func (s *Scanner) chr() rune {
	if s.eof() {
		return 0
	}
	return s.Data[s.pos]
}

func (s *Scanner) forward() {
	if s.Token != nil {
		s.Token.Text += string(s.chr())
	}

	s.pos++
	s.Pos.Col++

	if s.chr() == '\n' {
		s.Pos.Line++
		s.Pos.Col = 0
	}
}

func (s *Scanner) match(r rune) bool {
	if s.chr() == r {
		s.forward()
		return true
	}

	return false
}

func (s *Scanner) fmatch(f func(rune) bool) bool {
	if f(s.chr()) {
		s.forward()
		return true
	}

	return false
}

func (s *Scanner) expect(r rune) bool {
	if !s.match(r) {
		s.errorf("expected '%c', got '%c'", r, s.chr())
		return false
	}

	return true
}

func (s *Scanner) fexpect(f func(rune) bool) bool {
	if !s.fmatch(f) {
		s.errorf("unexpected character '%c'", s.chr())
		return false
	}

	return true
}

func (s *Scanner) eof() bool {
	return s.pos >= len(s.Data)
}

func isHex(r rune) bool {
	return unicode.IsDigit(r) || (r >= 'a' && r <= 'f') || (r >= 'A' && r <= 'F')
}

func isNameStart(r rune) bool {
	return unicode.IsLetter(r) || r == '_'
}

func isNamePart(r rune) bool {
	return isNameStart(r) || unicode.IsDigit(r) || r == '.'
}

func (s *Scanner) newTok() {
	s.Token = &Token{
		File: s.File,
		Pos:  s.Pos,
	}
}

func (s *Scanner) scanName() {
	s.Token.Kind = TokName
	for s.fmatch(isNamePart) {
	}
}

var charIntVal = []uint64{
	'0': 0,
	'1': 1,
	'2': 2,
	'3': 3,
	'4': 4,
	'5': 5,
	'6': 6,
	'7': 7,
	'8': 8,
	'9': 9,
	'a': 10, 'A': 10,
	'b': 11, 'B': 11,
	'c': 12, 'C': 12,
	'd': 13, 'D': 13,
	'e': 14, 'E': 14,
	'f': 15, 'F': 15,
}

func (s *Scanner) scanInt() {
	t := s.Token
	t.Kind = TokInt

	var base uint64 = 10

	if s.match('0') {
		switch {
		case s.match('x') || s.match('X'):
			base = 16

		case s.match('b') || s.match('B'):
			base = 2

		default:
			base = 8
		}
	}

	for {
		if s.match('_') {
			continue
		}

		digit := charIntVal[s.chr()]
		if digit == 0 && s.chr() != '0' {
			break
		}

		if digit >= base {
			s.errorf("digit '%c' out of range for base %d", s.chr(), base)
			digit = 0
		}

		if t.IntVal > math.MaxUint64-digit/base {
			s.errorf("integer literal overflow (%d)", t.IntVal)
			for s.fmatch(isHex) || s.match('_') {
			}

			t.IntVal = 0
			break
		}

		t.IntVal = t.IntVal*base + digit
		s.forward()
	}
}

var regVal = map[string]uint32{
	"%zero": 0, "%at": 1, "%v0": 2, "%v1": 3,
	"%a0": 4, "%a1": 5, "%a2": 6, "%a3": 7,
	"%t0": 8, "%t1": 9, "%t2": 10, "%t3": 11,
	"%t4": 12, "%t5": 13, "%t6": 14, "%t7": 15,
	"%s0": 16, "%s1": 17, "%s2": 18, "%s3": 19,
	"%s4": 20, "%s5": 21, "%s6": 22, "%s7": 23,
	"%t8": 24, "%t9": 25, "%k0": 26, "%k1": 27,
	"%gp": 28, "%sp": 29, "%fp": 30, "%ra": 31,
}

func (s *Scanner) scanRegister() {
	t := s.Token
	t.Kind = TokReg
	for s.fmatch(isNamePart) {
	}

	if IntVal, ok := regVal[t.Text]; ok {
		t.IntVal = uint64(IntVal)
	} else {
		s.errorf("unknown register '%s'", t.Text)
	}
}

func (s *Scanner) scanEscape() bool {
	t := s.Token
	switch {
	case s.match('n'):
		t.IntVal = '\n'
		t.StrVal += "\n"

	case s.match('r'):
		t.IntVal = '\r'
		t.StrVal += "\r"

	case s.match('t'):
		t.IntVal = '\t'
		t.StrVal += "\t"

	case s.match('\\'):
		t.IntVal = '\\'
		t.StrVal += "\\"

	case s.match('\''):
		t.IntVal = '\''
		t.StrVal += "'"

	case s.match('"'):
		t.IntVal = '"'
		t.StrVal += "\""

	case s.match('0'):
		t.IntVal = 0
		t.StrVal += "\\0"

	case s.match('x'): // hexadecimal
		s.fexpect(isHex)
		s.fmatch(isHex)

		esc := t.Text[3:]
		n, err := strconv.ParseInt(esc, 16, 0)
		if err != nil {
			t.StrVal += "\\" + esc
			return false
		}
		t.StrVal += string(n)
		t.IntVal = uint64(n)

	case s.match('u'): // unicode
		s.fexpect(isHex)
		s.fmatch(isHex)
		s.fmatch(isHex)
		s.fmatch(isHex)

		esc := t.Text[3:]
		n, _ := strconv.ParseInt(esc, 16, 0)
		u := make([]byte, 4)
		utf8.EncodeRune(u, rune(n))
		b := bytes.Trim(u, "\x00")
		t.StrVal += string(b) // remove NULL bytes
		t.IntVal = uint64(n)

	default:
		s.errorf("unexpected character in escape sequence '%c'", s.chr())
		return false
	}
	return true
}

func (s *Scanner) scanChar() {
	t := s.Token
	t.Kind = TokInt

	switch {
	case s.match('\\'):
		s.scanEscape()
		break

	default:
		t.IntVal = uint64(s.chr())
		s.forward()
	}
	s.expect('\'')
}

func (s *Scanner) scanString() {
	t := s.Token
	t.Kind = TokString

	for !s.eof() {
		switch {
		case s.eof() || s.match('\n'):
			s.errorf("unclosed string literal")
			return

		case s.fmatch(unicode.IsControl):
			s.errorf("illegal control character in string literal")

		case s.match('"'):
			return

		case s.match('\\'):
			s.scanEscape()

		default:
			t.StrVal += string(s.chr())
			s.forward()
		}
	}
}

func (s *Scanner) discardSpaces() {
	for s.fmatch(unicode.IsSpace) {
	}
}

func (s *Scanner) discardComment() {
	for !s.match('\n') {
		s.forward()
	}
}

/*
func (s *Scanner) discardBlockComment() bool {
	for depth := 1; depth > 0; {
		switch {
		case s.eof():
			s.errorf("unclosed block comment")
			return false

		case s.match('/') && s.match('*'):
			depth++

		case s.match('*') && s.match('/'):
			depth--

		default:
			s.forward()
		}
	}
	return true
}
*/

func (s *Scanner) Tokenize() {
	s.discardSpaces()

	s.newTok()
	t := s.Token

	switch {
	case s.eof():
		t.Kind = TokEOF
		return

	case s.fmatch(isNameStart):
		s.scanName()
		return

	case unicode.IsDigit(s.chr()):
		s.scanInt()
		return

	case s.match('%'):
		s.scanRegister()
		return

	case s.match('"'):
		s.scanString()
		return

	case s.match('\''):
		s.scanChar()
		return

	case s.match(';'):
		switch {
		case s.match(';'):
			s.discardComment()

		default:
			t.Kind = int(s.chr())
			s.forward()
			return
		}

		s.Tokenize()
		return

	case s.match('.'):
		s.scanName()
		t.Kind = TokDirective
		return

	default:
		t.Kind = int(s.chr())
		s.forward()
		return
	}
}

func safeRune(r rune) string {
	switch r {
	case '\n':
		return "\\n"

	case '\r':
		return "\\r"

	case '\t':
		return "\\t"

	case '\'':
		return "\\'"

	case '"':
		return "\\\""

	default:
		if unicode.IsPrint(r) {
			return string(r)
		}
		return fmt.Sprintf("\\%x", r)
	}
}
